docs/archive-data.md

```diff
@@ -0,0 +1,314 @@
+
+### Custom archiver FAQ
+
+**If my Archive method can automatically be retried by the caller, how can I record and access progress between retries?**
+
+Handle this using the `ArchiverOptions`. Here is an example:
+
+```go
+func(a * Archiver) Archive(ctx context.Context, URI string, request * ArchiveRequest, opts...ArchiveOption) error {
+    featureCatalog: = GetFeatureCatalog(opts...) // this function is defined in options.go
+    var progress progress
+    // Check if the feature for recording progress is enabled.
+    if featureCatalog.ProgressManager != nil {
+        if err: = featureCatalog.ProgressManager.LoadProgress(ctx, & prevProgress);
+        err != nil {
+            // log some error message and return error if needed.
+        }
+    }
+
+    // Your archiver implementation...
+
+    // Record current progress
+    if featureCatalog.ProgressManager != nil {
+        if err: = featureCatalog.ProgressManager.RecordProgress(ctx, progress);
+        err != nil {
+            // log some error message and return error if needed.
+        }
+    }
+}
+```
+
+**If my `Archive` method encounters an error which is non-retryable how do I indicate to the caller that should not retry?**
+
+```go
+func(a * Archiver) Archive(ctx context.Context, URI string, request * ArchiveRequest, opts...ArchiveOption) error {
+    featureCatalog: = GetFeatureCatalog(opts...) // this function is defined in options.go
+
+    err: = youArchiverImpl()
+  
+    if nonRetryableErr(err) {
+        if featureCatalog.NonRetryableError != nil {
+            return featureCatalog.NonRetryableError() // when the caller gets this error type back it will not retry anymore.
+        }
+    }
+}
+```
+
+**How does my history archiver implementation read history?**
+
+The archiver package provides a utility called [`HistoryIterator`](https://github.com/temporalio/temporal/blob/master/common/archiver/historyIterator.go) which is a wrapper of [`HistoryManager`](https://github.com/temporalio/temporal/blob/master/common/persistence/historyStore.go). `HistoryIterator` is more simple than the `HistoryManager`, which is available in the BootstrapContainer, so archiver implementations can choose to use it when reading Workflow histories. See the [historyIterator.go](https://github.com/temporalio/temporal/blob/master/common/persistence/historyStore.go) file for more details. Use the [filestore historyArchiver implementation](https://github.com/temporalio/temporal/tree/master/common/archiver/filestore) as an example.
+
+**Should my archiver define its own error types?**
+
+Each archiver is free to define and return its own errors. However many common errors which exist between archivers are already defined in [constants.go](https://github.com/temporalio/temporal/blob/master/common/archiver/constants.go).
+
+**Is there a generic query syntax for the visibility archiver?**
```

Comment:
There is no visibility archiver?

Reply:
There is - but it is not fully operational. And any archiver implementation needs to support it currently :)